File: java/utils/aoc_utils/geometry/Position2D.java
package aoc_utils.geometry;

import java.util.Objects;

public class Position2D<T> {
    public T x;
    public T y;

    public Position2D(T x, T y) {
        this.x = x;
        this.y = y;
    }

    public Position2D(Position2D<T> other) {
        this.x = other.x;
        this.y = other.y;
    }

    public boolean equals(Object obj) {
        if ( this == obj ) {
            return true;
        } else if ( !(obj instanceof Position2D<?> other) ) {
            return false;
        } else {
            return x == other.x && y == other.y;
        }
    }

    public int hashCode() {
        return Objects.hash(x, y);
    }
}

File: vbnet/utils/Geometry/Position2D.vb
Imports System

Namespace Geometry
	Public Structure Position2D(Of T)
		Implements IEquatable(Of Position2D(Of T))

		Public Property X As T
		Public Property Y As T

		Public Sub New(x As T, y As T)
			Me.X = x
			Me.Y = y
		End Sub

		Public Overloads Overrides Function Equals(obj As Object) As Boolean
			If TypeOf obj Is Position2D(Of T) Then
				Return Equals(CType(obj, Position2D(Of T)))
			End If
			Return False
		End Function

		Public Overloads Function Equals(other As Position2D(Of T)) As Boolean Implements IEquatable(Of Position2D(Of T)).Equals
			Return Me.X.Equals(other.X) AndAlso Me.Y.Equals(other.Y)
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return HashCode.Combine(Me.X, Me.Y)
		End Function

		Public Overrides Function ToString() As String
			Return $"Position2D({Me.X}, {Me.Y})"
		End Function
	End Structure
End Namespace
